lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -322,71 +326,82 @@ void getGuesses(std::vector<Inst *> &Guesses,
   // need two reserved per select instruction
   Inst *C2 = IC.getReservedConst();
   Comps.push_back(C2);
+  Inst *C3 = IC.getReservedConst();
+  Comps.push_back(C3);
   Inst *I3 = IC.getReservedInst();
   Comps.push_back(I3);
 
-  for (auto I = Comps.begin(); I != Comps.end(); ++I) {
-    if ((*I)->K == Inst::ReservedInst && (*I) != I1)
-      continue;
-    if ((*I)->K == Inst::ReservedConst && (*I) != C1)
-      continue;
-
-    for (auto J = Comps.begin(); J != Comps.end(); ++J) {
-      if ((*J)->K == Inst::ReservedInst && (*J) != I2)
+  for (auto Op : TernaryOperators) {
+    for (auto I : Comps) {
+      if (I->K == Inst::ReservedInst && I != I1)
         continue;
-      if ((*J)->K == Inst::ReservedConst && (*J) != C2)
+
+      if (I->K == Inst::ReservedConst && I != C1)
         continue;
 
-      // Prune (select cond, x, x)
-      if (I == J)
+      // (select i1, c, c)
+      // PRUNE: a select's control input should never be constant
+      if (Op == Inst::Select && I->K == Inst::ReservedConst)
         continue;
-      Inst *V1, *V2;
-      if ((*I)->K == Inst::ReservedConst) {
-        V1 = IC.createSynthesisConstant(Width, (*I)->SynthesisConstID);
-      } else if ((*I)->K == Inst::ReservedInst) {
-        V1 = IC.createHole(Width);
-      } else {
-        V1 = *I;
-      }
-      if ((*J)->K == Inst::ReservedConst) {
-        V2 = IC.createSynthesisConstant(Width, (*J)->SynthesisConstID);
-      } else if ((*J)->K == Inst::ReservedInst) {
-        V2 = IC.createHole(Width);
+
+      Inst *V1;
+      if (I->K == Inst::ReservedConst) {
+        V1 = IC.createSynthesisConstant(Width, I->SynthesisConstID);
+      } else if (I->K == Inst::ReservedInst) {
+        V1 = IC.createHole(Op == Inst::Select ? 1 : Width);
       } else {
-        V2 = *J;
+        V1 = I;
       }
 
-      if (V1->Width != V2->Width)
-        continue;
-
-      if (V1->Width != Width)
+      if (Op == Inst::Select && V1->Width != 1)
```

Comment:
Should there be a width check for the operators other than select?

Reply:
yup. added now.